### PR TITLE
[ENH] Implement scanners for performance enhancement on structured data

### PR DESCRIPTION
**Pull request by zzzzwj**
This PR implements the function on extracting all local patches from a image dataset, which is similar to ```sklearn.feature_extraction.image.extract_patches_2d``` but with more options, such as the kernel size, stride, padding size.

It is used in the original paper and designed to improve the performance of deep forest on image datasets. In addition, the scanner can also be used to pre-process the dataset and improve the performance of other classifiers.

After transformations, the training dataset will be in the shape ```(n_samples * n_patches_per_sample, kernel_size*kernel_size)```. Notice that since ```n_samples * n_patches_per_sample``` is typically very large, the data type of feature values should be cast into ```np.unit8``` after using the scanner.

Steps:
- [x] Add the PyTorch backend and GPU support
- [x] Add the Numpy backend
- [x] Add unit tests on the image scanner (Input: a toy image; Test: All patches should be exactly what we want)
- [ ] Add the documentation
- [ ] Verify the improvements after using the image scanner

### CODE REVIEW

**Comment by xuyxu** [comment]
@all-contributors please add @zzzzwj for code test

**Review by xuyxu — COMMENTED** [review]
LGTM.
